dex/txfee/oracle.go

```diff
@@ -0,0 +1,255 @@
+func (o *Oracle) RemoveFeeListener(uniqueID string) {
+	o.listenersMtx.Lock()
+	defer o.listenersMtx.Unlock()
+	feeChan, ok := o.listeners[uniqueID]
+	if !ok {
+		return
+	}
+
+	close(feeChan)
+	delete(o.listeners, uniqueID)
+}
+
+// calculateAverage calculates the average fee estimates and distributes the
+// result to all listeners. Returns indexes of newly reactivated sources that we
+// need to fetch fee estimate from.
+func (o *Oracle) calculateAverage() []int {
+	var reActivatedSourceIndexes []int
+	totalFeeEstimates := make(map[uint32]*FeeSourceCount)
+	for i := range o.sources {
+		source := o.sources[i]
+		if source.isDisabled() {
+			if source.checkIfSourceCanReactivate() {
+				reActivatedSourceIndexes = append(reActivatedSourceIndexes, i)
+			}
+			continue
+		}
+
+		source.mtx.Lock()
+		estimates := source.feeEstimates
+		source.mtx.Unlock()
+
+		for chainID, feeEstimate := range estimates {
+			if feeEstimate == 0 {
+				continue
+			}
+
+			if _, ok := totalFeeEstimates[chainID]; !ok {
+				totalFeeEstimates[chainID] = new(FeeSourceCount)
+			}
+
+			totalFeeEstimates[chainID].totalSource++
+			totalFeeEstimates[chainID].totalFee += feeEstimate
+		}
+	}
+
+	now := time.Now()
+	o.feeMtx.Lock()
+	broadCastTxFees := make(map[uint32]*Estimate, len(o.txFeeEstimates))
+	for chainID := range o.txFeeEstimates {
+		if rateInfo := totalFeeEstimates[chainID]; rateInfo != nil {
+			fee := rateInfo.totalFee / uint64(rateInfo.totalSource)
+			if fee > 0 {
+				o.txFeeEstimates[chainID].Value = fee
+				o.txFeeEstimates[chainID].LastUpdated = now
+				estimate := *o.txFeeEstimates[chainID]
+				broadCastTxFees[chainID] = &estimate
+			}
+		}
+	}
+	o.feeMtx.Unlock()
+
+	// Notify all listeners if we have rates to broadcast.
+	if len(broadCastTxFees) > 0 {
+		o.notifyListeners(broadCastTxFees)
+	}
+
+	return reActivatedSourceIndexes
+}
+
+func (o *Oracle) notifyListeners(feeEstimates map[uint32]*Estimate) {
+	o.listenersMtx.RLock()
+	defer o.listenersMtx.RUnlock()
+	for _, feeChan := range o.listeners {
+		feeChan <- feeEstimates
+	}
+}
+
+// Run starts the tx fee oracle and block until the provided context is
+// canceled.
+func (o *Oracle) Run(ctx context.Context, log dex.Logger) {
+	var wg sync.WaitGroup
+	for i := range o.sources {
+		source := o.sources[i]
+		if source.isDisabled() {
+			continue
+		}
+
+		o.fetchFromSource(ctx, log, source, &wg)
+
+		estimates, err := source.getFeeEstimate(ctx, log, o.chainIDs)
+		if err != nil {
+			if isAuthError(err) {
+				source.deactivate(false)
+				log.Errorf("%s has been deactivated and cannot be auto reactivated due to %v", source.name, err)
+			} else {
+				log.Errorf("%s.getFeeEstimate error: %v", source.name, err)
+			}
+			continue
+		}
+
+		source.mtx.Lock()
+		source.feeEstimates = estimates
+		source.mtx.Unlock()
+	}
+
+	o.calculateAverage()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		ticker := time.NewTicker(defaultFeeRefreshInterval + 10*time.Second)
```

Comment:
Wouldn't it make more sense for every `defaultFeeRefreshInterval`, we fetch all the sources, and then notify subscribers after they are all fetched. The way you have it currently, if one of the sources takes longer than 10 seconds to fetch, then the old value will be used in the calculation.

Reply:
> Wouldn't it make more sense for every defaultFeeRefreshInterval, we fetch all the sources, 

The refresh interval could differ for different sources, but considering that all sources we have atm refresh every `defaultFeeRefreshInterval`, your suggestion works better.

> The way you have it currently, if one of the sources takes longer than 10 seconds to fetch, then the old value will be used in the calculation.

Yes, It doesn't change much because the data expiry is longer than that.

Thanks for the review.